File: Adapters/Infoportal.Adapters.Elasticsearch/SearchContextMapping.cs
namespace Infoportal.Adapters.Elasticsearch;

public enum SearchContext
{
    All,
    StartCompany,
    Schema,
    Help
}

public static class SearchContextMapping
{
    // Maps content type aliases to search contexts.
    // Entries for content types that do not yet exist in Umbraco are harmless:
    // they simply never match until the content type is created and content is published.
    // Keep this list in sync with SearchContextMapping.ts on the frontend.
    private static readonly Dictionary<string, SearchContext> ContentTypeToContext = new(StringComparer.OrdinalIgnoreCase)
    {
        { "sectionArticlePage", SearchContext.StartCompany },
        { "subsidyPage", SearchContext.StartCompany },
        { "schemaPage", SearchContext.Schema },
        { "schemaCollectionPage", SearchContext.Schema },
        { "schemaAttachmentPage", SearchContext.Schema },
        { "helpQuestionPage", SearchContext.Help },
        { "helpProcessArticlePage", SearchContext.Help },
    };

    // Reverse lookup: context → list of content type aliases
    private static readonly Dictionary<string, string[]> ContextToContentTypes =
        ContentTypeToContext
            .GroupBy(kv => kv.Value)
            .ToDictionary(
                g => g.Key.ToString(),
                g => g.Select(kv => kv.Key).ToArray(),
                StringComparer.OrdinalIgnoreCase);

    /// <summary>
    /// Returns the content type aliases that belong to the given search context.
    /// Returns null for "All" or unknown contexts (no filtering needed).
    /// </summary>
    public static string[]? GetContentTypesForContext(string? context)
    {
        if (string.IsNullOrEmpty(context) || context.Equals(nameof(SearchContext.All), StringComparison.OrdinalIgnoreCase))
            return null;

        return ContextToContentTypes.TryGetValue(context, out var types) ? types : null;
    }

    /// <summary>
    /// Returns the search context for a given content type alias.
    /// Returns null if the content type is not mapped to any context.
    /// </summary>
    public static string? GetContextForContentType(string contentTypeAlias)
    {
        return ContentTypeToContext.TryGetValue(contentTypeAlias, out var context)
            ? context.ToString()
            : null;
    }

    /// <summary>
    /// Groups raw content type counts into search context counts.
    /// Filters out contexts with zero results.
    /// </summary>
    public static Dictionary<string, long> GroupByContext(IEnumerable<KeyValuePair<string, long>> contentTypeCounts)
    {
        var contextCounts = new Dictionary<string, long>(StringComparer.OrdinalIgnoreCase);

        foreach (var (contentType, count) in contentTypeCounts)
        {
            var context = GetContextForContentType(contentType);
            if (context == null) continue;

            if (contextCounts.ContainsKey(context))
                contextCounts[context] += count;
            else
                contextCounts[context] = count;
        }

        return contextCounts;
    }
}

File: apps/astro-infoportal/src/Services/Elasticsearch/SearchContextMapping.ts
const SearchContext = {
  All: "All",
  StartCompany: "StartCompany",
  Schema: "Schema",
  Help: "Help",
} as const;

// Entries for content types that do not yet exist in Umbraco are harmless:
// they simply never match until the content type is created and content is published.
// Keep this list in sync with SearchContextMapping.cs on the backend.
const contentTypeToContext: Record<string, string> = {
  sectionArticlePage: SearchContext.StartCompany,
  subsidyPage: SearchContext.StartCompany,
  schemaPage: SearchContext.Schema,
  schemaCollectionPage: SearchContext.Schema,
  schemaAttachmentPage: SearchContext.Schema,
  helpQuestionPage: SearchContext.Help,
  helpProcessArticlePage: SearchContext.Help,
};

const contextToContentTypes: Record<string, string[]> = {};
for (const [contentType, context] of Object.entries(contentTypeToContext)) {
  if (!contextToContentTypes[context]) {
    contextToContentTypes[context] = [];
  }
  contextToContentTypes[context].push(contentType);
}

/**
 * Returns the content type aliases that belong to the given search context.
 * Returns null for "All" or unknown contexts (no filtering needed).
 */
export function getContentTypesForContext(
  context: string | null | undefined,
): string[] | null {
  if (!context || context.toLowerCase() === "all") {
    return null;
  }
  for (const [key, types] of Object.entries(contextToContentTypes)) {
    if (key.toLowerCase() === context.toLowerCase()) {
      return types;
    }
  }
  return null;
}

/**
 * Returns the search context name for a given content type alias.
 * Returns null if the content type is not mapped.
 */
export function getContextForContentType(
  contentTypeAlias: string,
): string | null {
  return contentTypeToContext[contentTypeAlias] ?? null;
}

/**
 * Groups raw content type counts from ES aggregation buckets into search context counts.
 */
export function groupByContext(
  contentTypeCounts: Array<{ key: string; count: number }>,
): Record<string, number> {
  const contextCounts: Record<string, number> = {};

  for (const { key, count } of contentTypeCounts) {
    const context = getContextForContentType(key);
    if (!context) continue;

    contextCounts[context] = (contextCounts[context] ?? 0) + count;
  }

  return contextCounts;
}
